crates/ra_assists/src/handlers/add_function.rs

```diff
@@ -0,0 +1,716 @@
+// fn baz() -> Baz { Baz }
+// fn foo() {
+//      bar<|>("", baz());
+// }
+//
+// ```
+// ->
+// ```
+// struct Baz;
+// fn baz() -> Baz { Baz }
+// fn foo() {
+//      bar("", baz());
+// }
+//
+// fn bar(arg: &str, baz: Baz) {
+//     <|>todo!()
+// }
+//
+// ```
+pub(crate) fn add_function(ctx: AssistCtx) -> Option<Assist> {
+    let path: ast::Path = ctx.find_node_at_offset()?;
+
+    if ctx.sema.resolve_path(&path).is_some() {
+        // The function call already resolves, no need to add a function
+        return None;
+    }
+
+    if path.qualifier().is_some() {
+        return None;
+    }
+
+    let call: ast::CallExpr = ctx.find_node_at_offset()?;
+
+    if unresolved_path_is_not_call_expr(&path, &call)? {
+        return None;
+    }
+
+    let function_builder = FunctionBuilder::from_call(&ctx, &call)?;
+
+    ctx.add_assist(AssistId("add_function"), "Add function", |edit| {
+        edit.target(call.syntax().text_range());
+
+        let function_template = function_builder.render();
+        edit.set_cursor(function_template.cursor_offset);
+        edit.insert(function_template.insert_offset, function_template.fn_text);
+    })
+}
+
+/// Checks whether the path of the called function in `call_expr` equals `path`.
+///
+/// If it doesn't, the cursor is on an unresolved path of a non-function item in
+/// `call_expr` and we don't want to generate a function with that item's name.
+fn unresolved_path_is_not_call_expr(path: &ast::Path, call_expr: &ast::CallExpr) -> Option<bool> {
+    let call_path = call_expr.syntax().descendants().find_map(ast::Path::cast)?;
+    Some(call_path != *path)
+}
+
+struct FunctionTemplate {
+    insert_offset: TextUnit,
+    cursor_offset: TextUnit,
+    fn_text: String,
+}
+
+struct FunctionBuilder {
+    start_offset: TextUnit,
+    fn_name: String,
+    fn_generics: String,
+    fn_args: String,
+    indent: String,
+}
+
+impl FunctionBuilder {
+    fn from_call(ctx: &AssistCtx, call: &ast::CallExpr) -> Option<Self> {
+        let (start, indent) = next_space_for_fn(&call)?;
+        let fn_name = fn_name(&call)?;
+        let fn_generics = fn_generics(&call)?;
+        let fn_args = fn_args(ctx, &call)?;
+        let indent = if let Some(i) = &indent { i.to_string() } else { String::new() };
+        Some(Self { start_offset: start, fn_name, fn_generics, fn_args, indent })
+    }
+    fn render(&self) -> FunctionTemplate {
+        let mut fn_buf = String::with_capacity(128);
+        fn_buf.push_str("\n\n");
+        fn_buf.push_str(&self.indent);
+        fn_buf.push_str("fn ");
+        fn_buf.push_str(&self.fn_name);
+        fn_buf.push_str(&self.fn_generics);
+        fn_buf.push_str(&self.fn_args);
+        fn_buf.push_str(" {\n");
+        fn_buf.push_str(&self.indent);
+        fn_buf.push_str("    ");
+
+        // We take the offset here to put the cursor in front of the `todo` body
+        let offset = TextUnit::of_str(&fn_buf);
+
+        fn_buf.push_str("todo!()\n");
```

Comment:
I'm questioning the placeholder now. Intellij leaves the body empty.
If you just want to start typing something, the assist will always put something in your way first, which might get annoying.
Should I just leave the body empty?

Reply:
I might be asking for too much, but can we insert and select the placeholder, so that the user can overwrite it directly? :sweat_smile:.

Reply:
I'd like that, I haven't found a way to do that yet though. I'm looking at the `extend_selection` code for inspiration and I'm not finding anything I can use in the assist.

Reply:
I guess it would need some changes in the top-level assist handler, so far it looks like it only handles inserts and deletes.
Maybe we should move this to another PR as well

Reply:
> todo!() here breaks the tidy test.

```diff
-        fn_buf.push_str("todo!()\n");
+        fn_buf.push_str("tod");
+        fn_buf.push_str("o!()\n");
```

Reply:
unfortunately, we can't do this workaround in the doctest